src/config/validation.rs

```diff
@@ -189,6 +192,127 @@ pub fn check_outputs(config: &ConfigBuilder) -> Result<(), Vec<String>> {
     }
 }
 
+pub async fn check_buffer_preconditions(config: &Config) -> Result<(), Vec<String>> {
+    // We need to assert that Vector's data directory is located on a mountpoint that has enough
+    // capacity to allow all sinks with disk buffers configured to be able to use up to their
+    // maximum configured size without overrunning the total capacity.
+    //
+    // More subtly, we need to make sure we properly map a given buffer's data directory to the
+    // appropriate mountpoint, as it is technically possible that individual buffers could be on
+    // separate mountpoints.
+    //
+    // Notably, this does *not* cover other data usage by Vector on the same mountpoint because we
+    // don't always know the upper bound of that usage i.e. file checkpoint state.
+
+    // Grab all configured disk buffers, and if none are present, simply return early.
+    let global_data_dir = config.global.data_dir.clone();
+    let configured_disk_buffers = config
+        .sinks()
+        .flat_map(|(id, sink)| {
+            sink.buffer
+                .stages()
+                .iter()
+                .filter_map(|stage| stage.disk_usage(global_data_dir.clone(), id))
+        })
+        .collect::<Vec<_>>();
+
+    if configured_disk_buffers.is_empty() {
+        return Ok(());
+    }
+
+    // Now query all the mountpoints on the system, and get their total capacity. We also have to
+    // sort the mountpoints from longest to shortest so we can find the longest prefix match for
+    // each buffer data directory by simply iterating from beginning to end.
+    let mountpoints = heim::disk::partitions_physical()
+        .and_then(|partitions| {
+            partitions
+                .and_then(|partition| {
+                    let mountpoint_path = partition.mount_point().to_path_buf();
+                    heim::disk::usage(mountpoint_path.clone()).map(|usage| {
+                        usage.map(|usage| (mountpoint_path, usage.total().get::<byte>()))
+                    })
+                })
+                .try_collect::<IndexMap<_, _>>()
+        })
+        .await;
+
+    let mountpoints = match mountpoints {
+        Ok(mut mountpoints) => {
+            mountpoints.sort_by(|m1, _, m2, _| m2.cmp(m1));
+            mountpoints
+        }
+        Err(e) => {
+            warn!("Failed to query disk partitions. Cannot ensure that buffer size limits are within physical storage capacity limits.");
+            debug!(
+                "Encountered unexpected error during disk partition query: {}.",
+                e
+            );
```

Comment:
I assume this should be a single log before we merge this?

Reply:
I specifically made it two, mostly because I don't know how gnarly the error coming back from `heim` might be, and wanted to log it at a more hidden level than the warning.


But maybe it should just be a field on warning message. 🤔

Reply:
As long as we're in the "Should this actually be an error?" phase, it seems reasonable to raise it high enough to be normally seen - as long as we release it (and log it) with appropriate caveats.

Reply:
Yeah, I'll leave it as a warning -- because I'm still a bit unsure about making it a hard error without more real-world usage showing that it works in production without issues -- but I'll add the error as a `cause` field so that users can at least report back _why_ it failed and not need to rerun things with debug logging enabled to get assistance.

Reply:
I think that's a good balance 👍